### PR TITLE
[C++ API] AdaptiveLogSoftmaxWithLoss

### PR DESCRIPTION
**Pull request by mansoorcheema**
Implemented AdaptiveLogSoftmaxWithLoss and some tests for modules. Reference #25883

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/adaptive.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+  if (row_indices.numel() == 0) {
+    continue;
+  }
+
+  if (i == 0) {
+    gather_inds = gather_inds.index_copy(0, row_indices, target.index_select(0, row_indices));
+  }
+  else {
+    Tensor relative_target = target.index_select(0, row_indices) - low_idx;
```
[the remark this answers is not in the document]
Comment (mansoorcheema) [reply torch/csrc/api/src/nn/modules/adaptive.cpp]: @yf225 Yes this seems the normal way but the time I implemented it, aten did not have support for indexing using one or more than one dimensional indexes but only supported indexing using scalar tensors and for the return Tensor, which apparently should not be needed if the original Tensor is updated during the call. I also expected `.index_copy_()` to update the calling Tensor as other similar methods do but surprisingly it does not.

[the remark this answers is not in the document]
Comment (mansoorcheema) [reply torch/csrc/api/src/nn/modules/adaptive.cpp]: As mentioned in above comment, When I implemented this, there was no support for indexing tensors using multidimensional tensors, an doing so lead to exception with similar warning message.

File: torch/csrc/api/src/nn/modules/adaptive.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    const Tensor target_mask = (target >= low_idx) * (target < high_idx);
+    const Tensor row_indices = target_mask.nonzero().squeeze();
+
+  if (row_indices.numel() == 0) {
+    continue;
+  }
+
+  if (i == 0) {
+    gather_inds = gather_inds.index_copy(0, row_indices, target.index_select(0, row_indices));
```
[the remark this answers is not in the document]
Comment (mansoorcheema) [reply torch/csrc/api/src/nn/modules/adaptive.cpp]: @yf225 Yes this seems the normal way but the time I implemented it, aten did not have support for indexing using one or more than one dimensional indexes but only supported indexing using scalar tensors and for the return Tensor, which apparently should not be needed if the original Tensor is updated during the call. I also expected `.index_copy_()` to update the calling Tensor as other similar methods do but surprisingly it does not.

File: torch/csrc/api/src/nn/modules/adaptive.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    Tensor relative_target = target.index_select(0, row_indices) - low_idx;
+    Tensor input_subset = input.index_select(0, row_indices);
+
+    const Tensor cluster_output = tail[i - 1]->as<Sequential>()->forward(input_subset);
+    int64_t cluster_index = shortlist_size + i - 1;
+    gather_inds.index_fill_(0, row_indices, cluster_index);
+
+    const Tensor cluster_logprob = log_softmax(cluster_output, 1);
+    const Tensor local_logprob = cluster_logprob.gather(1, relative_target.unsqueeze(1));
+    output = output.index_copy_(0, row_indices, local_logprob.squeeze(1));
```
[the remark this answers is not in the document]
Comment (mansoorcheema) [reply torch/csrc/api/src/nn/modules/adaptive.cpp]: As mentioned in another comment `.index_copy_()` does not update the calling Tensor.

File: torch/csrc/api/src/nn/modules/adaptive.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mansoorcheema) [reply torch/csrc/api/src/nn/modules/adaptive.cpp]: I tried the latest version and still its not possible to index using 1D Tensors.

**Comment by mansoorcheema** [comment]
@yf225 Thanks for the feedback. I have committed the suggested changes except using target mask for indexing (multi-dimensional tensor indexing) which currently is not supported in the C++ API and  with the details mentioned inline.

**Comment by mansoorcheema** [comment]
@yf225 I have committed a few updates along with rebasing on latest code. Can you have a look?

**Review by yf225 — APPROVED** [review]
Thanks so much for the awesome work! Apologies for the delay as I was waiting for C++ tensor multi-dim indexing to be available, so that we don't need to use `narrow` / `index_select` :D

**Pull request closed without merging**